### ISSUE TITLE
Flawless Installation - Write Tests for `setup.ts`

### ISSUE THREAD

**Issue by palisadoes** [PalisadoesFoundation/talawa-api#1680]
**Is your feature request related to a problem? Please describe.**

1. We frequently add new configuration parameters to our code base.
2. The changes mean that the `setup.ts` file is always being modified.
3. Most PRs that modify this file introduce various types of failure.

PRs related to `setup.ts` require a high degree of manual verification by our reviewers and errors are frequent.

**Describe the solution you'd like**

Your job is to make the execution of the `setup.ts` file be:

1. foolproof
2. intuitive to use
3. easier and simpler to use

A tech novice or new contributor should be able to install the API without issue.

For the successful completion of this issue:

1. Write code to thoroughly test the `setup.ts` file without manually inputting responses to prompts, including:
    1.  all relevant functions, methods and classes where applicable.
    5. all user prompts
6. Ensure that running the tests is a part of our existing PR and Husky `git commit` workflows

This may mean making changes to the `setup.ts` file.

**Describe alternatives you've considered**

- N/A

**Approach to be followed (optional)**

This issue may require the refactoring of this script file to make it:

1. Easier to test
4. Easier to maintain

Testing must verify that the:

1. Default values for input are correct:
     1. When a parameter is already configured 
     1. When a parameter is not configured 
1. Functionality when parameters are missing from the configuration is maintained
1. Sequencing of the prompts is maintained
1. Configuration parameters that rely on any external APIs (eg. database, mail) must be tested.
    1. The tests don't hang when testing these APIs whether or not the API connectivity is validated
1. Preexisting comments in the test configuration file are maintained when the final configuration is saved
1. Final configuration isn't saved until the last question is answered. In other words, hitting CTRL-C at any point before the last question will not affect the original configuration.
1. Wording of the prompts in all scenarios is unchanged
1. Test database loads without issue
1. Previously correctly answered questions are not repeated
1. Correct values are echoed when:
    1. entered, 
    1. updated or 
    1. previously found in the configuration
1. Prompts correctly bypass sections that the user doesn't want to configure. 
    1. For example, there are multiple SMTP and Redis configuration parameters. 
        1. _Optional parameter sections_: If the person installing the API doesn't want to update SMTP parameters they should be skipped, otherwise prompted
        1.  _Mandatory parameter sections_: If the person installing the API doesn't want to update configured mandatory Redis parameters, they should be skipped, otherwise prompted
1. Input of mandatory parameters cannot be avoided if they are not previously configured

Plus any other tests that could avoid failure based on the PR comments for the issues below.

Other items to consider include:

1. The test file must be located in the existing `tests/` directory
1. The test file has good readability
1. Not making any changes to files in the `.github` directory

**Additional context**

These are issues where setup has been modified and in most cases have introduced more errors:

1. https://github.com/PalisadoesFoundation/talawa-api/issues/1625
1. https://github.com/PalisadoesFoundation/talawa-api/issues/1639
1. https://github.com/PalisadoesFoundation/talawa-api/issues/1606
1. https://github.com/PalisadoesFoundation/talawa-api/issues/1550
1. https://github.com/PalisadoesFoundation/talawa-api/issues/1531
1. https://github.com/PalisadoesFoundation/talawa-api/issues/1526
1. https://github.com/PalisadoesFoundation/talawa-api/issues/1524
1. https://github.com/PalisadoesFoundation/talawa-api/issues/1443
1. https://github.com/PalisadoesFoundation/talawa-api/issues/1337
1. https://github.com/PalisadoesFoundation/talawa-api/issues/1313

**Potential internship candidates**
Please read this if you are planning to apply for a Palisadoes Foundation internship 
- https://github.com/PalisadoesFoundation/talawa/issues/359

**Comment by varshith257** [comment]
I would like to work on this.

**Comment by AVtheking** [comment]
I would like to work on this issue

**Comment by LakshyaSatpal** [comment]
I can work on this

**Comment by varshith257** [comment]
@palisadoes  There are many changes will occur to existing setup.ts in PR #1632 that included node env,redis connection,super admin mail etc....

Done with this task for existing setup.ts: 
Refactoring of this script file to make it:

Easier to test
Easier to maintain

whether I should wait the PR #1632 to be merged or would I continue with existing setup.ts

**Comment by palisadoes** [comment]
You'll need to wait. That PR should be completed soon

**Comment by varshith257** [comment]
@palisadoes  Waiting for the update.

**Comment by palisadoes** [comment]
The PR was merged yesterday

**Comment by LakshyaSatpal** [comment]
Please assign this to me.

**Comment by LakshyaSatpal** [comment]
The `main` function inside `setup.ts` contains 300 lines of code, making it hard to test.
I'm thinking of dividing it into atomic functions.

**Comment by palisadoes** [comment]
Yes. Make the setup code more readable and compliant with best practices

**Comment by palisadoes** [comment]
unassigning due to inactivity

**Comment by adi790uu** [comment]
@palisadoes would like to solve this.

**Comment by adi790uu** [comment]
I'm unassigning myself from this since as I've already been assigned two other issues. I'll revisit this if it's still not solved @Cioppolo14

**Comment by Anubhav-2003** [comment]
@palisadoes Respected sir, 

I would like to work on this issue. My latest PR #1815 brings a significant change to the setup.ts related to encryption of emails. I am familiar with the file. Also the setup at the moment has some unused parts. I would be writing tests, make the setup code more intuitive, as well incorporate the new features into the setup.

Thank You.

**Comment by Cioppolo14** [comment]
@Anubhav-2003 I have assigned you. For future requests related to assignments, please feel free to tag me instead of palisadoes.

**Comment by Anubhav-2003** [comment]
@Cioppolo14 Ok ma'am. Thank You.

**Comment by palisadoes** [comment]
You'll need to fix this warning that comes up as part of the setup process

![image](https://github.com/PalisadoesFoundation/talawa-api/assets/16875803/26155f9c-2501-4a6e-8e1f-5ca27b1deaa4)

**Comment by Anubhav-2003** [comment]
@palisadoes ok sir.

**Comment by Anubhav-2003** [comment]
@palisadoes Respecteds sir,

 I just wanted to ask if there was any update regarding the review of the PR #1815 , as I am unable to fully work on this issue until that PR merges, as that PR introduces a new feature with a substantial in the setup.ts. Which I would need to inculcate in this issue as well.

Thank You.

**Comment by palisadoes** [comment]
1. Please work on this issue without considering #1815 
2. It will be easy to review and test. I'll do review as an exception as this really needs to be done
3. Update #1815 after the PR for this issue is merged.

**Comment by Anubhav-2003** [comment]
@palisadoes ok sir.

**Comment by Anubhav-2003** [comment]
Respected sir, 

I  just wanted to give an update on the issue, as it has been almost a week. I am working on fully automated test cases, to covering all the functionalities of the setup.ts. I have refactored and changed the setup file to make it more testable. I have also changed the order of the prompts for Image size, and import sample data as requested by you. There are over a 1000 lines of code in setup.ts So it is taking some time. I will raise a PR as soon as it is done.

Thank You.

**Comment by palisadoes** [comment]
Thanks for the update. 

It's a very important part of our code base. Take care with it

**Comment by Anubhav-2003** [comment]
Absolutely sir.
